extensions/ql-vscode/src/data-extensions-editor/auto-model-usages-query.ts

```diff
@@ -0,0 +1,129 @@
+import { CancellationTokenSource } from "vscode";
+import { join } from "path";
+import { runQuery } from "./external-api-usage-query";
+import { CodeQLCliServer } from "../codeql-cli/cli";
+import { QueryRunner } from "../query-server";
+import { DatabaseItem } from "../databases/local-databases";
+import { interpretResultsSarif } from "../query-results";
+import { ProgressCallback } from "../common/vscode/progress";
+
+type Options = {
+  cliServer: Pick<
+    CodeQLCliServer,
+    "resolveDatabase" | "resolveQlpacks" | "interpretBqrsSarif"
+  >;
+  queryRunner: Pick<QueryRunner, "createQueryRun" | "logger">;
+  databaseItem: Pick<
+    DatabaseItem,
+    | "contents"
+    | "databaseUri"
+    | "language"
+    | "sourceArchive"
+    | "getSourceLocationPrefix"
+  >;
+  queryStorageDir: string;
+
+  progress: ProgressCallback;
+};
+
+export async function getAutoModelUsages({
+  cliServer,
+  queryRunner,
+  databaseItem,
+  queryStorageDir,
+  progress,
+}: Options): Promise<Record<string, string[]>> {
+  const maxStep = 1500;
+
+  const cancellationTokenSource = new CancellationTokenSource();
+
+  const queryResult = await runQuery("usagesQuery", {
+    cliServer,
+    queryRunner,
+    queryStorageDir,
+    databaseItem,
+    progress: (update) =>
+      progress({
+        maxStep,
+        step: update.step,
+        message: update.message,
+      }),
+    token: cancellationTokenSource.token,
+  });
+  if (!queryResult) {
+    throw new Error("Query failed");
+  }
+
+  progress({
+    maxStep,
+    step: 1100,
+    message: "Retrieving source locatin prefix",
+  });
+
+  const sourceLocationPrefix = await databaseItem.getSourceLocationPrefix(
+    cliServer,
+  );
+  const sourceArchiveUri = databaseItem.sourceArchive;
+  const sourceInfo =
+    sourceArchiveUri === undefined
+      ? undefined
+      : {
+          sourceArchive: sourceArchiveUri.fsPath,
+          sourceLocationPrefix,
+        };
+
+  progress({
+    maxStep,
+    step: 1200,
+    message: "Interpreting results",
+  });
+
+  const sarif = await interpretResultsSarif(
+    cliServer,
+    {
+      kind: "problem",
+      id: "usage",
+    },
+    {
+      resultsPath: queryResult.outputDir.bqrsPath,
+      interpretedResultsPath: join(
+        queryStorageDir,
+        "interpreted-results.sarif",
+      ),
+    },
+    sourceInfo,
+    ["--sarif-add-snippets"],
+  );
+
+  progress({
+    maxStep,
+    step: 1400,
+    message: "Parsing results",
+  });
+
+  const snippets: Record<string, string[]> = {};
```

Comment:
What do you think about moving the logic that extracts the snippets into `sarif-utils`? It'll make this function a bit smaller and it'll make it easier to test this aspect too.

Reply:
I don't think it really makes sense to move this into `sarif-utils`. This code is not really reusable since it's very specific to the SARIF file that is being produced by this exact query. Usually, you'd not want to create a map out of it since the message is the alert message.

For just the part `result.locations?.[0]?.physicalLocation?.contextRegion?.snippet?.text`, this is also not reusable. Usually, you would have multiple locations since the path contains multiple nodes.

I can extract the grouping into a separate function, but then we'd be exporting it only for testing. This is probably fine, but what do you think?

Reply:
Ah yeah, it's not really re-usable, you're right. Let's leave as is.